src/journal/JournalMetadata.cc

```diff
@@ -776,10 +785,16 @@ void JournalMetadata::handle_refresh_complete(C_Refresh *refresh, int r) {
       lderr(m_cct) << "failed to locate client: " << m_client_id << dendl;
       r = -ENOENT;
     }
+
+    assert(m_refreshes_in_progress > 0);
+    --m_refreshes_in_progress;
+    if (m_refreshes_in_progress == 0) {
+      std::swap(refresh_ctxs, m_refresh_ctxs);
+    }
```

Comment:
@dillaman I suppose it should be outside of `if (r == 0)` block.

Observing rbd_mirror.sh test failure, on "disable mirroring / delete non-primary image", due to image replayer shut_down hangs on `m_remote_journaler->shut_down`, when refresh mutable metadata returns r=-2 (for test3 or test5 image). See e.g. [1] but it is reproduced locally too.

[1] http://pulpito.ceph.com/trociny-2018-02-14_14:24:46-rbd-wip-mgolub-testing-distro-basic-smithi/

Reply:
D'oh -- definitely